File: src/rating_bot/jobs.py
"""The two scheduled jobs: new applications to review, and deadline reminders."""

import asyncio
import logging
from collections import defaultdict

from aiogram import Bot
from aiogram.enums import ParseMode

from rating_bot.config import Config
from rating_bot.dateutil import DT, now
from rating_bot.db import Database
from rating_bot.helpers import (
    DEFAULT_HOST,
    format_applications,
    get_application_form,
    get_batches,
)
from rating_bot.ratingutil import (
    get_applications,
    get_info,
    info_is_bad,
    tourn_info_to_reminders,
)
from rating_bot.subscriptions import review_link

logger = logging.getLogger("rating_bot")


async def send(bot: Bot, chat_id: int, text: str) -> None:
    for batch in get_batches(text):
        try:
            await bot.send_message(chat_id, batch, parse_mode=ParseMode.HTML)
        except Exception as e:
            logger.error(f"exception {type(e)} {e} while messaging {chat_id}")


async def check_applications(
    bot: Bot, db: Database, config: Config, chat_ids_whitelist=None
) -> None:
    logger.debug("running regular job...")
    pending_writes = []
    user_to_message = defaultdict(list)
    user_to_subscriptions = defaultdict(list)
    tournament_applications = {}

    for tournament in db.tournaments():
        for chat_id, subscription in tournament.subscribers.items():
            if subscription.get("r"):
                user_to_subscriptions[chat_id].append(
                    (tournament.id, tournament.name)
                )
        try:
            info = await asyncio.to_thread(get_info, tournament.id)
            if not chat_ids_whitelist and info_is_bad(info):
                logger.debug(f"adding request for removal of tourn_id {tournament.id}")
                pending_writes.append((db.delete_tournament, (tournament.id,)))
                continue
            if DT(info["dateEnd"]) > DT(now()):
                await _collect_new_applications(
                    db,
                    tournament,
                    chat_ids_whitelist,
                    pending_writes,
                    user_to_message,
                    tournament_applications,
                )
        except Exception as e:
            logger.error(
                f"error while processing tourn_id {tournament.id}: {type(e)} {e}"
            )
            continue

    logger.debug(f"total {len(user_to_message)} chats will be messaged")
    for chat_id in user_to_message:
        if chat_ids_whitelist and chat_id not in chat_ids_whitelist:
            continue
        try:
            texts = _messages_for_chat(
                db, chat_id, user_to_message, user_to_subscriptions, tournament_applications
            )
        except Exception as e:
            logger.error(
                f"error while trying to process messages for {chat_id}: {type(e)} {e}"
            )
            continue
        final_text = "\n\n".join(texts)
        if final_text:
            logger.debug(f"sending messages to {chat_id}")
            await send(bot, chat_id, final_text)

    if not chat_ids_whitelist:
        _apply(pending_writes)


async def _collect_new_applications(
    db, tournament, chat_ids_whitelist, pending_writes, user_to_message, tournament_applications
):
    fresh = await asyncio.to_thread(get_applications, tournament.id)
    if fresh is None:
        logger.error(
            f"could not get applications for tourn_id {tournament.id}, skipping"
        )
        return
    added = set(fresh) - set(tournament.applications)
    removed = set(tournament.applications) - set(fresh)
    tournament_applications[(tournament.id, tournament.name)] = fresh
    if (added or removed) and not chat_ids_whitelist:
        logger.debug(
            f"adding request for updating data for tourn_id {tournament.id}:"
            f" was {tournament.applications}, became {fresh}"
        )
        pending_writes.append((db.set_applications, (tournament.id, fresh)))
    if not added:
        return
    logger.debug(
        f"adding requests for sending messages for tourn_id {tournament.id}"
        f" for chat_ids {','.join(str(x) for x in tournament.subscribers)}"
    )
    for chat_id, subscription in tournament.subscribers.items():
        if (
            chat_ids_whitelist and chat_id not in chat_ids_whitelist
        ) or not subscription.get("r"):
            continue
        host = db.prefs(chat_id).get("host") or DEFAULT_HOST
        text = (
            f"Для турнира <b>{tournament.id} {tournament.name}</b> есть {len(fresh)}"
            f" {get_application_form(len(fresh))}. {review_link(host, tournament.id)}\n\n"
            + format_applications(fresh, host)
        )
        user_to_message[chat_id].append((tournament.id, text))
        logger.debug(f"added message for {chat_id} about {tournament.id}")


def _messages_for_chat(
    db, chat_id, user_to_message, user_to_subscriptions, tournament_applications
):
    reported = user_to_message[chat_id]
    texts = [text for _, text in reported]
    reported_ids = {tournament_id for tournament_id, _ in reported}
    host = db.prefs(chat_id).get("host") or DEFAULT_HOST
    others = [t for t in user_to_subscriptions[chat_id] if t[0] not in reported_ids]
    for tournament_id, name in sorted(others):
        applications = tournament_applications.get((tournament_id, name))
        if applications:
            texts.append(
                f"Ранее нерассмотренные заявки на турнир <b>{tournament_id} {name}</b>."
                f" {review_link(host, tournament_id)}\n\n"
                + format_applications(applications, host)
            )
    return texts


async def make_reminders(bot: Bot, db: Database, config: Config) -> None:
    logger.debug("running regular job...")
    pending_writes = []
    user_to_message = defaultdict(list)

    for tournament in db.tournaments():
        logger.info(f"processing tournament {tournament.id}...")
        info = await asyncio.to_thread(get_info, tournament.id)
        if info_is_bad(info):
            logger.debug(f"adding request for removal of tourn_id {tournament.id}")
            pending_writes.append((db.delete_tournament, (tournament.id,)))
            continue
        if (DT(now()).dt - DT(info["dateEnd"]).dt).days >= 30:
            logger.debug(f"adding request for removal of tourn_id {tournament.id}")
            pending_writes.append((db.delete_tournament, (tournament.id,)))
            continue
        logger.debug(f"getting reminders for tournament {tournament.id}")
        try:
            reminders = await asyncio.to_thread(tourn_info_to_reminders, info, DT(now()))
        except Exception as e:
            logger.debug(
                f"exception {type(e)} {e} while trying to get reminders for {tournament.id}"
            )
            continue
        if not reminders:
            continue
        logger.debug(f"got reminders for tournament {tournament.id}")
        for chat_id, subscription in tournament.subscribers.items():
            wanted = {key for key in subscription if subscription[key]} & set(reminders)
            if not wanted:
                continue
            seen = []
            for key in wanted:
                if reminders[key] not in seen:
                    seen.append(reminders[key])
            user_to_message[chat_id].append("\n\n".join(seen))

    logger.info(f"got messages for {len(user_to_message)} chats")
    for chat_id, messages in user_to_message.items():
        logger.debug(f"processing messages to {chat_id}")
        text = "\n\n".join(messages)
        host = db.prefs(chat_id).get("host") or DEFAULT_HOST
        if host != DEFAULT_HOST:
            text = text.replace(DEFAULT_HOST, host)
        await send(bot, chat_id, text)

    _apply(pending_writes)


async def check_applications_debug(bot: Bot, db: Database, config: Config) -> None:
    await check_applications(bot, db, config, chat_ids_whitelist=config.admins)


async def test_job(bot: Bot, db: Database, config: Config) -> None:
    for chat_id in config.admins:
        await send(bot, chat_id, "<b>test</b>")


def _apply(pending_writes) -> None:
    if not pending_writes:
        return
    logger.debug("committing requests")
    for write, args in pending_writes:
        try:
            write(*args)
        except Exception as e:
            logger.debug(f"exception while trying to commit {args}: {type(e)} {e}")
    logger.debug("end of committing requests")

File: internal/tg/jobs.go
package tg

import (
	"context"
	"fmt"
	"log/slog"
	"slices"
	"sort"
	"strings"
	"time"

	"code.pecheny.me/pecheny/rating_chgk_info_helper_bot/internal/dates"
	"code.pecheny.me/pecheny/rating_chgk_info_helper_bot/internal/rating"
	"code.pecheny.me/pecheny/rating_chgk_info_helper_bot/internal/store"
	"code.pecheny.me/pecheny/rating_chgk_info_helper_bot/internal/text"
)

// forgetAfter is how long a finished tournament stays watched. Past it there is
// nothing left to rule on and nobody left to remind.
const forgetAfter = 30 * 24 * time.Hour

// news is one tournament's fresh applications, already written out for one
// chat. watched is a tournament that chat follows but heard nothing about.
type newsItem struct {
	tournamentID int
	body         string
}

type watchedItem struct {
	tournamentID int
	name         string
}

// CheckApplications tells every subscriber about applications they have not
// reviewed yet. It runs every two hours.
func (b *Bot) CheckApplications(ctx context.Context) {
	b.checkApplications(ctx, nil)
}

// checkApplicationsForAdmins is the same sweep sent only to the admins, which
// is how a change gets tried against live data without messaging anyone.
func (b *Bot) checkApplicationsForAdmins(ctx context.Context) {
	b.checkApplications(ctx, b.config.Admins)
}

func (b *Bot) checkApplications(ctx context.Context, only map[int64]bool) {
	slog.Debug("running the applications job")
	dryRun := only != nil

	tournaments, err := b.db.Tournaments()
	if err != nil {
		slog.Error("reading tournaments", "err", err)
		return
	}

	var stale []int
	news := map[int64][]newsItem{}
	watching := map[int64][]watchedItem{}
	current := map[int]map[string]store.Application{}

	for _, tournament := range tournaments {
		for chatID, subscription := range tournament.Subscribers {
			if subscription["r"] != 0 {
				watching[chatID] = append(watching[chatID],
					watchedItem{tournament.ID, tournament.Name})
			}
		}
		info, err := b.rating.Info(tournament.ID)
		if err != nil {
			slog.Error("processing tournament", "tournament", tournament.ID, "err", err)
			continue
		}
		if !dryRun && info.IsBad() {
			slog.Debug("tournament is gone from the site", "tournament", tournament.ID)
			stale = append(stale, tournament.ID)
			continue
		}
		end, err := dates.Parse(info.DateEnd)
		if err != nil {
			slog.Error("reading the end date", "tournament", tournament.ID, "err", err)
			continue
		}
		if !end.After(dates.Now()) {
			continue
		}
		b.collectNewApplications(tournament, only, news, current)
	}

	slog.Debug("chats to message", "count", len(news))
	for chatID := range news {
		if dryRun && !only[chatID] {
			continue
		}
		body := strings.Join(b.messagesFor(chatID, news, watching, current), "\n\n")
		if body == "" {
			continue
		}
		slog.Debug("sending applications digest", "chat_id", chatID)
		b.send(ctx, chatID, body, true)
	}

	if dryRun {
		return
	}
	for _, tournamentID := range stale {
		if err := b.db.DeleteTournament(tournamentID); err != nil {
			slog.Error("forgetting tournament", "tournament", tournamentID, "err", err)
		}
	}
}

func (b *Bot) collectNewApplications(
	tournament store.Tournament,
	only map[int64]bool,
	news map[int64][]newsItem,
	current map[int]map[string]store.Application,
) {
	fresh, err := b.rating.Applications(tournament.ID)
	if err != nil {
		slog.Error("could not get applications, skipping", "tournament", tournament.ID, "err", err)
		return
	}
	stored := toStoredApplications(fresh)
	current[tournament.ID] = stored

	if only == nil && differs(tournament.Applications, stored) {
		slog.Debug("application list changed", "tournament", tournament.ID,
			"was", len(tournament.Applications), "became", len(stored))
		if err := b.db.SetApplications(tournament.ID, stored); err != nil {
			slog.Error("saving applications", "tournament", tournament.ID, "err", err)
		}
	}
	if !hasNew(tournament.Applications, stored) {
		return
	}

	for chatID, subscription := range tournament.Subscribers {
		if (only != nil && !only[chatID]) || subscription["r"] == 0 {
			continue
		}
		host := b.db.HostOf(chatID, text.DefaultHost)
		body := fmt.Sprintf("Для турнира <b>%d %s</b> есть %d %s. %s\n\n%s",
			tournament.ID, tournament.Name, len(stored), text.ApplicationForm(len(stored)),
			text.ReviewLink(host, tournament.ID), formatApplications(stored, host))
		news[chatID] = append(news[chatID], newsItem{tournament.ID, body})
	}
}

// messagesFor is what one chat is told: the tournaments with new applications,
// then a nudge about the ones it still has not got to.
func (b *Bot) messagesFor(
	chatID int64,
	news map[int64][]newsItem,
	watching map[int64][]watchedItem,
	current map[int]map[string]store.Application,
) []string {
	var out []string
	told := map[int]bool{}
	for _, item := range news[chatID] {
		out = append(out, item.body)
		told[item.tournamentID] = true
	}

	host := b.db.HostOf(chatID, text.DefaultHost)
	others := make([]watchedItem, 0, len(watching[chatID]))
	for _, item := range watching[chatID] {
		if !told[item.tournamentID] {
			others = append(others, item)
		}
	}
	sort.Slice(others, func(i, j int) bool { return others[i].tournamentID < others[j].tournamentID })

	for _, item := range others {
		applications := current[item.tournamentID]
		if len(applications) == 0 {
			continue
		}
		out = append(out, fmt.Sprintf(
			"Ранее нерассмотренные заявки на турнир <b>%d %s</b>. %s\n\n%s",
			item.tournamentID, item.name, text.ReviewLink(host, item.tournamentID),
			formatApplications(applications, host)))
	}
	return out
}

// MakeReminders warns subscribers about the controversials and appeals
// deadlines. It runs once a day.
func (b *Bot) MakeReminders(ctx context.Context) {
	slog.Debug("running the reminders job")
	tournaments, err := b.db.Tournaments()
	if err != nil {
		slog.Error("reading tournaments", "err", err)
		return
	}

	var stale []int
	perChat := map[int64][]string{}

	for _, tournament := range tournaments {
		slog.Info("processing tournament", "tournament", tournament.ID)
		info, err := b.rating.Info(tournament.ID)
		if err != nil {
			slog.Error("getting tournament info", "tournament", tournament.ID, "err", err)
			continue
		}
		if info.IsBad() {
			stale = append(stale, tournament.ID)
			continue
		}
		end, err := dates.Parse(info.DateEnd)
		if err != nil {
			slog.Error("reading the end date", "tournament", tournament.ID, "err", err)
			continue
		}
		if dates.Now().Sub(end) >= forgetAfter {
			slog.Debug("tournament is long over", "tournament", tournament.ID)
			stale = append(stale, tournament.ID)
			continue
		}
		reminders, err := b.rating.RemindersFor(info, dates.Now())
		if err != nil {
			slog.Debug("could not build reminders", "tournament", tournament.ID, "err", err)
			continue
		}
		for chatID, subscription := range tournament.Subscribers {
			if wanted := wantedReminders(subscription, reminders); wanted != "" {
				perChat[chatID] = append(perChat[chatID], wanted)
			}
		}
	}

	slog.Info("chats to remind", "count", len(perChat))
	for chatID, messages := range perChat {
		body := strings.Join(messages, "\n\n")
		if host := b.db.HostOf(chatID, text.DefaultHost); host != text.DefaultHost {
			body = strings.ReplaceAll(body, text.DefaultHost, host)
		}
		b.send(ctx, chatID, body, true)
	}

	for _, tournamentID := range stale {
		if err := b.db.DeleteTournament(tournamentID); err != nil {
			slog.Error("forgetting tournament", "tournament", tournamentID, "err", err)
		}
	}
}

// wantedReminders picks the reminders this subscription asked for, without
// repeating one that covers two kinds at once.
func wantedReminders(subscription store.Subscription, reminders rating.Reminders) string {
	var chosen []string
	for _, kind := range []string{"i", "a"} {
		if subscription[kind] == 0 {
			continue
		}
		reminder, ok := reminders[kind]
		if !ok || reminder == "" {
			continue
		}
		if !slices.Contains(chosen, reminder) {
			chosen = append(chosen, reminder)
		}
	}
	return strings.Join(chosen, "\n\n")
}

func differs(before, after map[string]store.Application) bool {
	if len(before) != len(after) {
		return true
	}
	for id := range after {
		if _, known := before[id]; !known {
			return true
		}
	}
	return false
}

func hasNew(before, after map[string]store.Application) bool {
	for id := range after {
		if _, known := before[id]; !known {
			return true
		}
	}
	return false
}

func formatApplications(applications map[string]store.Application, host string) string {
	reps := make([]string, 0, len(applications))
	for _, application := range applications {
		reps = append(reps, application.Rep)
	}
	return text.FormatApplications(reps, host)
}
